File: LibraryWebApp.AuthorService/Application/Services/AuthorService.cs
using LibraryWebApp.AuthorService.Domain.Entities;
using LibraryWebApp.AuthorService.Domain.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace LibraryWebApp.AuthorService.Application.Services
{
    public class AuthorService : IAuthorService
    {
        private readonly IUnitOfWork _unitOfWork;

        public AuthorService(IUnitOfWork unitOfWork) 
        {
            _unitOfWork = unitOfWork;
        }

        public int GetAuthorId(string firstName, string lastName)
        {
            var author = _unitOfWork.Authors.GetAll()
                    .FirstOrDefault(a => a.FirstName!.Equals(firstName)
                                     && a.LastName!.Equals(lastName));

            return author!.Id;
        }

        public IEnumerable<Author> GetAllAuthors(int pageNumber, int pageSize)
        {
            return _unitOfWork.Authors.GetAll()
                .Skip((pageNumber - 1) * pageSize)
                .Include(a => a.Books)
                .Take(pageSize);
        }

        public void AddAuthor(Author author)
        {
            _unitOfWork.Authors.Create(author);
            _unitOfWork.Save();
        }

        public Author GetAuthor(int id)
        {
            var author = _unitOfWork.Authors.Get(a => a.Id == id);

            return author;
        }

        public void UpdateAuthor(Author author)
        {
            _unitOfWork.Authors.Update(author);
            _unitOfWork.Save();
        }

        public void DeleteAuthor(Author author)
        {
            _unitOfWork.Authors.Delete(author);
            _unitOfWork.Save();
        }

        public IEnumerable<Book> GetAllBooksByAuthor(Author author, int pageNumber, int pageSize)
        {
            return _unitOfWork.Books.GetAll()
                .Include(b => b.Author)
                .Where(b => b.Author == author)
                .Skip((pageNumber - 1) * pageSize)
                .Take(pageSize);
        }
    }
}

File: AngularProject/src/app/services/author.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AuthorDTO } from '../models/author-dto.model';
import { BookDTO } from '../models/book-dto.model';

@Injectable({
  providedIn: 'root'
})
export class AuthorService {
  private apiUrl = 'https://localhost:7000/author-service/Author';

  constructor(private http: HttpClient) { }

  getAuthorId(firstName: string, lastName: string): Observable<number> {
    const params = { firstName, lastName };
    return this.http.get<number>(`${this.apiUrl}/by-name`, { params });
  }

  getAllAuthors(pageNumber: number, pageSize: number): Observable<AuthorDTO[]> {
    return this.http.get<AuthorDTO[]>(`${this.apiUrl}?pageNumber=${pageNumber}&pageSize=${pageSize}`, {
    });
  }

  getAuthor(id: number): Observable<AuthorDTO> {
    return this.http.get<AuthorDTO>(`${this.apiUrl}/${id}`, {
    });
  }

  addAuthor(authorDto: AuthorDTO): Observable<void> {
    return this.http.post<void>(this.apiUrl, authorDto, {
      headers: this.getAuthHeaders()
    });
  }

  updateAuthor(id: number, authorDto: AuthorDTO): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}`, authorDto, {
      headers: this.getAuthHeaders()
    });
  }

  deleteAuthor(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`, {
      headers: this.getAuthHeaders()
    });
  }

  getAllBooksByAuthor(id: number, pageNumber: number = 1, pageSize: number = 10): Observable<BookDTO[]> {
    return this.http.get<BookDTO[]>(`${this.apiUrl}/${id}/books?pageNumber=${pageNumber}&pageSize=${pageSize}`, {
    });
  }

  private getAuthHeaders(): { [header: string]: string } {
    const token = localStorage.getItem('accessToken');
    return {
      'Authorization': `Bearer ${token}`
    };
  }
}
